rfcs/text/0352-graph-type-system.md

```diff
@@ -0,0 +1,2659 @@
+
+```json
+{
+  "name": "someProperty",
+  "oneOf": [{ "$ref": "data-type/number" }, { "$ref": "data-type/text" }]
+}
+```
+
+More specifically, it's required to allow `oneOf` with more variants than _just Data Types_ because:
+
+```json
+{
+  "someProperty": [{ "bar": 2.0 }, "foo"]
+}
+```
+
+Requires something like
+
+```json
+{
+  "name": "someProperty",
+  "oneOf": [
+    {
+      "type": "object",
+      "properties": {
+        "bar": { "$ref": "property-type/bar" }
+      }
+    },
+    { "$ref": "data-type/text" }
+  ]
+}
+```
+
+### Including an `Empty List` Data Type
+
+Similarly, the `Empty List` Data Type is required to be able to programmatically map unknown JSON data into the system. Given an incoming entity like:
+
+```json
+{
+    "foo": [
+        [],
+        [2, 6, 10, ...],
+        [{...}, {...}, ...]
+    ]
+}
+```
+
+There isn't a good assumption to make about the type of the first array. As such, including `Empty List` makes it possible to treat this as a value without an inner type, and a user will be forced to retype it to be able to modify it.
+
+### Including an `Object` Data Type
+
+The inclusion of the `Object` Data Type is less strongly supported. At the moment it seems like it's easy to include in implementations (as datastores will already need to be able to represent JSON-like structures), so instead the discussion at the moment is around looking for a reason to exclude it.
+
+There's a risk that it could be utilized instead of Property Type objects, which means the inner data won't be accessible through tooling designed around the Type System. However at the moment it seems a convenient inclusion to allow users to prototype and quickly throw in semi-structured data as black boxes, later on transitioning to Data Types. As the type is very constrained in the current proposal, with it not being possible to add further JSON schema keywords, etc. the risk that users will misuse it seems low.
+
+# Prior art
+
+[prior-art]: #prior-art
+
+As mentioned in a few sections, this design basically defines a way for communities to build a compound ontology. As such it's worth mentioning some of the alternative knowledge representation models to provide points of discussion, and to invite members of their communities to give thoughts. There are a few specific technologies that we've taken particular interest in while designing the RFC.
+
+> Note: Some planning has been made around **varying**-levels of compatibility with these technologies, although in-depth designs for their interfacing have not been written up yet.
+
+## Data Description Formats
+
+- [RDF](https://www.w3.org/RDF/)
+- [RDFS](https://www.w3.org/TR/rdf-schema/)
+- [OWL](https://www.w3.org/OWL/)
+
+## Ontologies
+
+- [Schema.org](https://schema.org/)
+- [DBPedia](https://dbpedia.org/)
+
+# Unresolved questions
+
+[unresolved-questions]: #unresolved-questions
+
+1.  Are there further impacts on block schemas?
+1.  Do we permit spaces in type names?
```

Comment:
Why is this a question? Aren't they just display names?

Reply:
> I'm inclined to think we should treat them as plain text and spaces are fine.

That's also my instinct

> perhaps there are cases where EAs would benefit from these not having spaces?

Perhaps, but I can't think of a reason - perhaps we can get rid of this unresolved question as we are all agreeing we like spaces in display names, and can't come up with a reason why spaces are problematic?

Reply:
Addressed the immediate decision in https://github.com/blockprotocol/blockprotocol/pull/352/commits/0bd6547994b1a205523473d5230633554d181c9a